app/com/linkedin/drelephant/spark/heuristics/ConfigurationParametersHeuristic.scala

```diff
@@ -0,0 +1,485 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.linkedin.drelephant.spark.heuristics
+
+import com.linkedin.drelephant.analysis._
+import com.linkedin.drelephant.configurations.heuristic.HeuristicConfigurationData
+import com.linkedin.drelephant.spark.data.SparkApplicationData
+import com.linkedin.drelephant.spark.fetchers.statusapiv1.{ExecutorSummary, TaskDataImpl}
+
+import scala.collection.JavaConverters
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.commons.io.FileUtils
+
+import org.apache.spark.network.util.JavaUtils
+
+/**
+  * A heuristic for recommending configuration parameter values, based on metrics from the application run.
+  * @param heuristicConfigurationData
+  */
+class ConfigurationParametersHeuristic(private val heuristicConfigurationData: HeuristicConfigurationData)
+  extends Heuristic[SparkApplicationData] {
+
+  import JavaConverters._
+  import ConfigurationParametersHeuristic._
+  import ConfigurationHeuristicsConstants._
+
+  override def getHeuristicConfData(): HeuristicConfigurationData = heuristicConfigurationData
+
+  override def apply(data: SparkApplicationData): HeuristicResult = {
+
+    lazy val appConfigurationProperties: Map[String, String] =
+      data.appConfigurationProperties
+
+    val evaluator = new Evaluator(this, data)
+
+    // add current configuration parameter values, and recommended parameter values to the result.
+    var resultDetails = ArrayBuffer(
+      new HeuristicResultDetails(CURRENT_SPARK_EXECUTOR_MEMORY,
+        bytesToString(evaluator.sparkExecutorMemory)),
+      new HeuristicResultDetails(CURRENT_SPARK_DRIVER_MEMORY,
+        bytesToString(evaluator.sparkDriverMemory)),
+      new HeuristicResultDetails(CURRENT_SPARK_EXECUTOR_CORES, evaluator.sparkExecutorCores.toString),
+      new HeuristicResultDetails(CURRENT_SPARK_DRIVER_CORES, evaluator.sparkDriverCores.toString),
+      new HeuristicResultDetails(CURRENT_SPARK_MEMORY_FRACTION, evaluator.sparkMemoryFraction.toString))
+    evaluator.sparkExecutorInstances.foreach { numExecutors =>
+      resultDetails += new HeuristicResultDetails(CURRENT_SPARK_EXECUTOR_INSTANCES, numExecutors.toString)
+    }
+    evaluator.sparkExecutorMemoryOverhead.foreach { memOverhead =>
+      resultDetails += new HeuristicResultDetails(CURRENT_SPARK_EXECUTOR_MEMORY_OVERHEAD,
+        bytesToString(memOverhead))
+    }
+    evaluator.sparkDriverMemoryOverhead.foreach { memOverhead =>
+      resultDetails += new HeuristicResultDetails(CURRENT_SPARK_DRIVER_MEMORY_OVERHEAD,
+        bytesToString(memOverhead))
+    }
+
+    resultDetails ++= Seq(
```

Comment:
From the point of view of separation of concern, I would suggest that recommended parameters should be calculated within TuneIn code.

Reply:
From earlier discussions with @mkumar1984 and @pralabhkumar , there was a preference for keeping these recommendations as a separate heuristic from the current TuneIn recommendations, so that they could be evaluated and compared, and then later integrated with TuneIn.

Reply:
So do you mean that we will monitor these heuristics for the time being and later move this to TuneIn once we are comfortable with the recommendations?

Reply:
The current plan is to monitor by querying the db, and not expose the recommendations to users, in case having 2 sets of recommendations would be confusing for users. Integration with TuneIn would be done later.